### PR TITLE
Argnums in Jax transforms

### PR DESCRIPTION
**Pull request by rmoyard**
**Context:**

https://github.com/PennyLaneAI/pennylane/pull/3697 changes the definition of trainable parameters with Jax.

**Description of the Change:**

In this PR we allwo gradient transforms and hessian transforms to set `argnums` like we do in Jax. We do that by wrapping the trainable parameters with a JVPTracer.

**Benefits:**

Those transforms are compatible with Jax.

**Possible Drawbacks:**

**Related GitHub Issues:**

**TODO:**

Add tests for the new return types after https://github.com/PennyLaneAI/pennylane/pull/3776 is merged.

### CODE REVIEW

File: pennylane/fourier/qnode_spectrum.py [hunk 1]
```diff
@@ -390,7 +390,7 @@ def wrapper(*args, **kwargs):
             qnode.interface = qml.math.get_interface(*args, *list(kwargs.values()))
 
         jac_fn = qml.transforms.classical_jacobian(
-            qnode, argnum=argnum, expand_fn=qml.transforms.expand_multipar
+            qnode, argnums=argnum, expand_fn=qml.transforms.expand_multipar
```
Comment (rmoyard) [inline pennylane/fourier/qnode_spectrum.py:393]:
```suggestion
            qnode, argnum=argnum, expand_fn=qml.transforms.expand_multipar
```

File: tests/transforms/test_classical_jacobian.py [hunk 2]
```diff
@@ -544,3 +443,175 @@ def test_torch_not_trainable_only(diff_method, interface):
...
+    ),
+    [
+        np.eye(len(x)),
+        np.eye(len(w)),
+        tuple(np.eye(len(x) + len(w)).reshape((2, len(x), len(x) + len(w))).transpose([0, 2, 1])),
+    ],
+]
+
+
+class TestJax:
```
Comment (dwierichs) [inline tests/transforms/test_classical_jacobian.py:523]: Docstring

File: tests/transforms/test_classical_jacobian.py [hunk 3]
```diff
@@ -544,3 +443,175 @@ def test_torch_not_trainable_only(diff_method, interface):
...
+    interfaces = ["jax"]
```
Comment (dwierichs) [inline tests/transforms/test_classical_jacobian.py:524]: Is this in preparation for something? I saw that this is used all over this file, but it seems unnecessary?

File: tests/transforms/test_classical_jacobian.py [hunk 4]
```diff
@@ -544,3 +443,175 @@ def test_torch_not_trainable_only(diff_method, interface):
...
+
+    @pytest.mark.jax
+    @pytest.mark.parametrize("diff_method", ["backprop", "parameter-shift"])
```
Comment (dwierichs) [inline tests/transforms/test_classical_jacobian.py:527]:
```suggestion
```

File: tests/transforms/test_classical_jacobian.py [hunk 5]
```diff
@@ -544,3 +443,175 @@ def test_torch_not_trainable_only(diff_method, interface):
...
+        ),
+    ),
+    [
+        np.eye(len(x)),
+        np.eye(len(w)),
+        tuple(np.eye(len(x) + len(w)).reshape((2, len(x), len(x) + len(w))).transpose([0, 2, 1])),
+    ],
+]
+
+
```
Comment (dwierichs) [inline tests/transforms/test_classical_jacobian.py:522]:
```suggestion

@pytest.mark.jax
@pytest.mark.parametrize("diff_method", ["backprop", "parameter-shift"])
```

File: tests/transforms/test_classical_jacobian.py [hunk 6]
```diff
@@ -544,3 +443,175 @@ def test_torch_not_trainable_only(diff_method, interface):
...
+        args = tuple((jnp.array(arg) for arg in args))
+        # JAX behaviour: argnum=None yields only the Jacobian with respect to the first arg.
+        expected_jac = expected_jac[0]
+        dev = qml.device("default.qubit", wires=2)
+        qnode = qml.QNode(circuit, dev, interface=interface, diff_method=diff_method)
+        jac = classical_jacobian(qnode)(*args)
+        assert np.allclose(jac, expected_jac)
+
+    @pytest.mark.jax
+    @pytest.mark.parametrize("diff_method", ["backprop", "parameter-shift"])
```
Comment (dwierichs) [inline tests/transforms/test_classical_jacobian.py:543]:
```suggestion
```

File: tests/transforms/test_classical_jacobian.py [hunk 7]
```diff
@@ -544,3 +443,175 @@ def test_torch_not_trainable_only(diff_method, interface):
...
+        args = tuple((jnp.array(arg) for arg in args))
+        dev = qml.device("default.qubit", wires=2)
+        qnode = qml.QNode(circuit, dev, interface=interface, diff_method=diff_method)
+        jac = classical_jacobian(qnode, argnum=argnum)(*args)
+        expected_jac = (expected_jac[argnum[0]],)
+        assert len(jac) == 1
+        assert np.allclose(jac[0], expected_jac[0])
+
+    @pytest.mark.jax
+    @pytest.mark.parametrize("diff_method", ["backprop", "parameter-shift"])
```
Comment (dwierichs) [inline tests/transforms/test_classical_jacobian.py:565]:
```suggestion
```

File: tests/transforms/test_classical_jacobian.py [hunk 8]
```diff
@@ -544,3 +443,175 @@ def test_torch_not_trainable_only(diff_method, interface):
...
+
+        args = tuple((jnp.array(arg) for arg in args))
+        dev = qml.device("default.qubit", wires=2)
+        qnode = qml.QNode(circuit, dev, interface=interface, diff_method=diff_method)
+        jac = classical_jacobian(qnode, argnum=argnum)(*args)
+        expected_jac = expected_jac[argnum]
+        assert np.allclose(jac, expected_jac)
+
+    @pytest.mark.jax
+    @pytest.mark.parametrize("diff_method", ["backprop", "parameter-shift"])
```
Comment (dwierichs) [inline tests/transforms/test_classical_jacobian.py:584]:
```suggestion
```

File: tests/transforms/test_classical_jacobian.py [hunk 9]
```diff
@@ -544,3 +443,175 @@ def test_torch_not_trainable_only(diff_method, interface):
...
+        if len(argnum) > 1:
+            expected_jac = expected_jac[-1]
+        print(argnum)
+        print(jac, expected_jac)
+        assert len(jac) == len(expected_jac)
+        for _jac, _expected_jac in zip(jac, expected_jac):
+            assert np.allclose(_jac, _expected_jac)
+
+    @pytest.mark.jax
+    @pytest.mark.parametrize("diff_method", ["backprop", "parameter-shift"])
```
Comment (dwierichs) [inline tests/transforms/test_classical_jacobian.py:609]:
```suggestion
```

File: pennylane/gradients/parameter_shift_hessian.py [hunk 10]
```diff
@@ -650,16 +650,6 @@ def param_shift_hessian(tape, argnum=None, diagonal_shifts=None, off_diagonal_sh
...
-
-    # For Jax the argnum defines the tape trainable parameters (it does not work for an array argnum).
-    if (
-        argnum is not None
-        and not isinstance(argnum, np.ndarray)
-        and not tape.trainable_params
-        and qml.math.get_interface(*tape.data) in ["jax", "jax-jit"]
-    ):
-        tape.trainable_params = argnum
-
```
Comment (dwierichs) [inline pennylane/gradients/parameter_shift_hessian.py:662]: There is a second place in this file where a parallel deletion should be made, no?
See [here](https://github.com/PennyLaneAI/pennylane/blob/a92677ffd977f0fd622f4bedd59095534e8c64fb/pennylane/gradients/parameter_shift_hessian.py#L880)

**Review by dwierichs — COMMENTED** [review]
Thanks for this, this brings #3697 into even nicer shape, I think! :100: :)
See my detailed comments in #3697 .
Mostly left minor suggestions/comments here.

File: pennylane/gradients/parameter_shift_hessian.py (hunk 10 above)
Comment (rmoyard) [reply pennylane/gradients/parameter_shift_hessian.py:662]: Thanks good catch :+1:

File: pennylane/transforms/metric_tensor.py [hunk 11]
```diff
@@ -325,6 +325,8 @@ def qnode_execution_wrapper(self, qnode, targs, tkwargs):
     mt_fn = self.default_qnode_wrapper(qnode, targs, tkwargs)
 
     def _expand_fn(tape):
+        # Currently does not support it
```
[the remark this answers is not in the document]
Comment (rmoyard) [reply pennylane/transforms/metric_tensor.py:328]: Yes now I will add it now that it is merged :+1:

File: tests/gradients/test_finite_difference.py [hunk 12]
```diff
@@ -816,3 +816,122 @@ def cost_fn(x):
...
+        )(x, y)
+        res_expected = jax.hessian(circuit, argnums=argnums)(x, y)
+
+        assert len(res) == len(res_expected)
+
+        if len(argnums) != 1:
+            res = res[0]
+
+        for r, r_e in zip(res, res_expected[0]):
+            tol = 10e-1
```
[the remark this answers is not in the document]
Comment (rmoyard) [reply tests/gradients/test_finite_difference.py:928]: I will improve it :+1:

File: pennylane/gradients/gradient_transform.py [hunk 13]
```diff
@@ -303,16 +307,23 @@ def jacobian_wrapper(*args, **kwargs):
...
+                )(*args, **kwargs)
+            else:
+                cjac = cjac_fn(*args, **kwargs)
 
             if isinstance(cjac, tuple):
                 # Classical processing of multiple arguments is present. Return qjac @ cjac.
                 jacs = tuple(
                     qml.math.tensordot(qjac, c, [[-1], [0]]) for c in cjac if c is not None
                 )
+                if len(jacs) == 1:
```
Comment (rmoyard) [inline pennylane/gradients/gradient_transform.py:325]: @dwierichs I added this change to match our convention where we do not wrap the jacobian in a tuple if there is only one, hence the change in the torch change.

File: tests/gradients/test_gradient_transform.py [hunk 14]
```diff
@@ -642,7 +642,7 @@ def circuit(x):
 
         x_ = -0.654
         x = torch.tensor(x_, dtype=torch.float64, requires_grad=True)
-        res = circuit(x)[0]
+        res = circuit(x)
```
[the remark this answers is not in the document]
Comment (rmoyard) [reply tests/gradients/test_gradient_transform.py:645]: see the comment above,

File: tests/transforms/test_classical_jacobian.py (hunk 3 above)
Comment (rmoyard) [reply tests/transforms/test_classical_jacobian.py:526]: We might add `auto` or `jax-python`, `jax-jit`

File: pennylane/math/multi_dispatch.py [hunk 15]
```diff
@@ -880,6 +880,29 @@ def detach(tensor, like=None):
     return tensor
 
 
+def jax_argnums_to_tape_trainable(qnode, argnums, expand_fn, args, kwargs):
+    """This functions gets the tape parameters from the QNode construction given some argnums. The tape parameters
+    are transformed to JVPTracer if they are from trainable parameters to mimic the behavior of Jax and  alos mark
+    them trainable.
```
[the remark this answers is not in the document]
Comment (rmoyard) [reply pennylane/math/multi_dispatch.py]: Sorry for the confusion, I updated it!

File: pennylane/math/multi_dispatch.py [hunk 16]
```diff
@@ -880,6 +880,29 @@ def detach(tensor, like=None):
...
+
+    args_jvp = [
+        jax.ad.JVPTracer(trace, arg, jax.numpy.zeros(1)) if i in argnums else arg
+        for i, arg in enumerate(args)
+    ]
+
+    qnode.construct(args_jvp, kwargs)
+    tape = qnode.qtape
+    tape = expand_fn(tape)
+    params_tape = tape.get_parameters(trainable_only=False)
```
[the remark this answers is not in the document]
Comment (rmoyard) [reply pennylane/math/multi_dispatch.py:902]: Yes it would be great to have something less "manual" but unfortunately if the transform is not directly taken from Jax I did not find a better way. Let me know if you have some ideas!

I will post a benchmark of this function.

**Comment by rmoyard** [comment]
@dwierichs Here is a benchmark for a middle size example:

```
import timeit

def jvp_trainable_time():
    SETUP_CODE = '''
import pennylane as qml
import pennylane.numpy as np

import jax
import jax.numpy as jnp

qml.enable_return()
dev = qml.device("default.qubit", wires=2)

@qml.qnode(dev, cache=False)
def circuit(params, params2, params3, params4, params5, params6):
    qml.RY(params[0], wires=0)
    qml.RX(params[1], wires=1)
    qml.RY(params[2], wires=0)
    qml.RX(params[3], wires=1)
    qml.RY(params[4], wires=0)
    qml.RX(params[5], wires=1)
    qml.RY(params2[0], wires=0)
    qml.RX(params2[1], wires=1)
    qml.RY(params2[2], wires=0)
    qml.RX(params2[3], wires=1)
    qml.RY(params2[4], wires=0)
    qml.RX(params2[5], wires=1)
    qml.RY(params3, wires=0)
    qml.RX(params4, wires=1)
    qml.RY(params5, wires=0)
    qml.RX(params6, wires=1)
    qml.CNOT(wires=[0, 1])
    return qml.expval(qml.PauliZ(0)), qml.expval(qml.PauliY(1))

a = jax.numpy.array([0.1, 0.2, 0.3, 0.4, 0.5, 0.6])
b = jax.numpy.array([0.6, 0.5, 0.4, 0.3, 0.2, 0.1])
c = jax.numpy.array(0.1)
d = jax.numpy.array(0.2)
e = jax.numpy.array(0.3)
f = jax.numpy.array(0.4)
argnums = [0, 2, 4, 5]

def fun(tape):
    return tape
'''

    TEST_CODE = '''

qml.math.jax_argnums_to_tape_trainable(circuit, argnums, fun, (a, b, c, d, e, f), {})'''

    # timeit.repeat statement
    times = timeit.repeat(setup=SETUP_CODE,
                          stmt=TEST_CODE,
                          repeat=100,
                          number=1)
    print('Average time: {}'.format(np.mean(times)))
    return np.mean(times)

jvp_trainable_time()
```
```
**Average time: 0.00793067308550235**
```

**Review by dwierichs — APPROVED** [review]
Thanks for the explanation and addressing everything.
Only the metric tensor left to integrate, but this could be left to another PR if needed.

File: pennylane/gradients/gradient_transform.py [hunk 17]
```diff
@@ -278,18 +278,22 @@ def default_qnode_wrapper(self, qnode, targs, tkwargs):
         # to take into account that classical processing may be present
         # inside the QNode.
         hybrid = tkwargs.pop("hybrid", self.hybrid)
-
-        old_interface = qnode.interface
```
Comment (timmysilv) [inline pennylane/gradients/gradient_transform.py:282]: why are you removing the old_interface stuff in this PR?

Comment (rmoyard) [reply pennylane/gradients/gradient_transform.py:282]: This was introduced by me and it was wrong to do it there because those are the transform arg not the qnode arg...
